File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.java
import java.util.*;
import java.io.*;

class Solution {
    
    boolean vi[] = new boolean[51];
    int answer = 100;
    public int solution(String begin, String target, String[] words) {
        dfs(begin, target, words, 0);
        if(answer == 100) answer = 0;
        return answer;
    }
    
    public void dfs(String begin, String target, String[] words, int tmp){
        if(begin.equals(target)){
            if(answer >= tmp) answer = tmp;
            return;
        }
        
        for(int i = 0; i < words.length; i++){
            int cnt = 0;
            for(int j = 0; j < words[i].length(); j++){
                if(begin.charAt(j) != words[i].charAt(j)) cnt++;
                if(cnt == 2) break;
            }
            if(cnt == 1){
                if(!vi[i]){
                    vi[i] = true;
                    dfs(words[i], target, words, tmp+1);
                    vi[i] = false;
                }
            }
        }
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <climits>
using namespace std;
bool visited[51];
int answer = INT_MAX;


void dfs(string cur, string target, int count, vector<string> words) {
    if (target == cur) {
        answer = min(answer, count);
        return;
    }
    
    for (int i = 0; i < words.size(); i++) {
        int charCount = 0;
        for (int j = 0; j < words[i].size(); j++) {
            if (words[i][j] != cur[j]) charCount++;
            if (charCount >= 2) break;
        }
        if (charCount == 1) {
            if (!visited[i]) {
                 visited[i] = true;
                 dfs(words[i], target, count + 1, words);
                 visited[i] = false;
            }
        }
    }
}
int solution(string begin, string target, vector<string> words) {
    
    dfs(begin, target, 0, words);
    return answer == INT_MAX ? 0 : answer;
}